pkg/nsx/services/ipaddressallocation/ipaddressallocation.go

```diff
@@ -0,0 +1,197 @@
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	ipAddressAllocationService := &IPAddressAllocationService{Service: service, VPCService: vpcService}
+	ipAddressAllocationService.ipAddressAllocationStore = &IPAddressAllocationStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPAddressAllocationCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBindingType(),
+	}}
+
+	tags := []model.Tag{
+		{Scope: String(common.TagScopeIPAddressAllocationCRUID)},
+	}
+	go ipAddressAllocationService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPAddressAllocation, tags, ipAddressAllocationService.ipAddressAllocationStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return ipAddressAllocationService, err
+	}
+	return ipAddressAllocationService, nil
+}
+
+func (service *IPAddressAllocationService) CreateOrUpdateIPAddressAllocation(obj *v1alpha1.IPAddressAllocation) (bool, error) {
+	nsxIPAddressAllocation, err := service.BuildIPAddressAllocation(obj)
+	if err != nil {
+		return false, err
+	}
+	existingIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
+		return false, err
+	}
+	log.V(1).Info("existing ipaddressallocation", "ipaddressallocation", existingIPAddressAllocation)
+	_ = common.CompareResource(IpAddressAllocationToComparable(existingIPAddressAllocation),
+		IpAddressAllocationToComparable(nsxIPAddressAllocation))
+
+	//if !ipAddressAllocationUpdated{
+	//	return false, nil
+	//}
+
+	if err := service.Apply(nsxIPAddressAllocation); err != nil {
+		return false, err
+	}
+
+	createdIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
+		return false, err
+	}
+	cidr := createdIPAddressAllocation.AllocationIps
+	if cidr == nil {
+		return false, fmt.Errorf("ipaddressallocation %s didn't realize available cidr", obj.UID)
+	}
+	obj.Status.CIDR = *cidr
+	return true, nil
+}
+
+func (service *IPAddressAllocationService) Apply(nsxIPAddressAllocation *model.VpcIpAddressAllocation) error {
+	ns := service.GetIPAddressAllocationNamespace(nsxIPAddressAllocation)
+	var err error
+	VPCInfo := service.VPCService.ListVPCInfo(ns)
+	if len(VPCInfo) == 0 {
+		err = util.NoEffectiveOption{Desc: "no valid org and project for ipaddressallocation"}
+		return err
+	}
+	err = service.NSXClient.IPAddressAllocationClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, VPCInfo[0].ID, *nsxIPAddressAllocation.Id, *nsxIPAddressAllocation)
+	err = util.NSXApiError(err)
+	if err != nil {
+		// not return err, try to get it from nsx, in case if cidr not realized at the first time
+		// support realize multiple times
+		log.Error(err, "patch failed, try to get it from nsx")
```

Comment:
Please add the object info for troubleshooting

Reply:
ok